File: problems/链表/206.反转链表.cpp
#include <iostream>
#include <vector>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
};

ListNode* listToLinkedList(vector<int> lst) {
    ListNode* dummy = new ListNode(0);
    ListNode* cur = dummy;
    for (int val : lst) {
        cur->next = new ListNode(val);
        cur = cur->next;
    }
    return dummy->next;
}

vector<int> linkedListToList(ListNode* node) {
    vector<int> result;
    while (node) {
        result.push_back(node->val);
        node = node->next;
    }
    return result;
}

ListNode* reverseList(ListNode* head) {
    ListNode* prev = nullptr;
    ListNode* cur = head;
    while (cur) {
        ListNode* nextNode = cur->next;
        cur->next = prev;
        prev = cur;
        cur = nextNode;
    }
    return prev;
}

void printList(ListNode* node) {
    cout << "[";
    while (node) {
        cout << node->val;
        if (node->next) cout << ",";
        node = node->next;
    }
    cout << "]";
}

int main() {
    // 测试用例1: 1->2->3->4->5 反转为 5->4->3->2->1
    ListNode* head1 = listToLinkedList({1, 2, 3, 4, 5});
    ListNode* result1 = reverseList(head1);
    cout << "测试1 - 反转 [1,2,3,4,5]:" << endl;
    cout << "  输出: ";
    printList(result1);
    cout << endl;

    // 测试用例2: 1->2 反转为 2->1
    ListNode* head2 = listToLinkedList({1, 2});
    ListNode* result2 = reverseList(head2);
    cout << "\n测试2 - 反转 [1,2]:" << endl;
    cout << "  输出: ";
    printList(result2);
    cout << endl;

    // 测试用例3: 单节点
    ListNode* head3 = listToLinkedList({1});
    ListNode* result3 = reverseList(head3);
    cout << "\n测试3 - 反转 [1]:" << endl;
    cout << "  输出: ";
    printList(result3);
    cout << endl;

    return 0;
}

File: problems/链表/206.反转链表.py
#!/usr/bin/env python3
"""
206. 反转链表 (Reverse Linked List)
题目: 给你单链表的头节点 head，请你反转链表，并返回反转后的链表。
"""

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def reverse_list(head: ListNode) -> ListNode:
    prev = None
    cur = head
    while cur:
        next_node = cur.next
        cur.next = prev
        prev = cur
        cur = next_node
    return prev

def list_to_linkedlist(lst):
    dummy = ListNode(0)
    cur = dummy
    for val in lst:
        cur.next = ListNode(val)
        cur = cur.next
    return dummy.next

def linkedlist_to_list(node):
    result = []
    while node:
        result.append(node.val)
        node = node.next
    return result

if __name__ == "__main__":
    # 测试用例1: 1->2->3->4->5 反转为 5->4->3->2->1
    head1 = list_to_linkedlist([1, 2, 3, 4, 5])
    result1 = reverse_list(head1)
    print("测试1 - 反转 [1,2,3,4,5]:")
    print(f"  输出: {linkedlist_to_list(result1)}")  # [5, 4, 3, 2, 1]

    # 测试用例2: 1->2 反转为 2->1
    head2 = list_to_linkedlist([1, 2])
    result2 = reverse_list(head2)
    print("\n测试2 - 反转 [1,2]:")
    print(f"  输出: {linkedlist_to_list(result2)}")  # [2, 1]

    # 测试用例3: 单节点 [] 反转后仍为 []
    head3 = list_to_linkedlist([1])
    result3 = reverse_list(head3)
    print("\n测试3 - 反转 [1]:")
    print(f"  输出: {linkedlist_to_list(result3)}")  # [1]
